File: client/web/src/features/vault/api/vault-api.ts
import { apiClient } from "@/lib/api-client";
import type { VaultSyncResponse } from "@/types/api-types";

export function fetchVault(): Promise<VaultSyncResponse> {
  return apiClient.get<VaultSyncResponse>("/vault");
}

export interface VaultUpsertResponse {
  id: string;
  entryVersion: number;
  updatedAt: string;
}

export function upsertEntry(
  entryId: string,
  entryPayload: string,
  entryVersion: number,
): Promise<VaultUpsertResponse> {
  return apiClient.put<VaultUpsertResponse>(`/vault/${entryId}`, {
    entryPayload,
    entryVersion,
  });
}

export function deleteEntry(entryId: string): Promise<void> {
  return apiClient.delete<void>(`/vault/${entryId}`);
}

export function reorderEntries(entryIds: string[]): Promise<void> {
  return apiClient.post<void>("/vault/reorder", { entryIds });
}

File: client/android/core/src/main/java/io/github/pstanar/pstotp/core/api/VaultApi.kt
package io.github.pstanar.pstotp.core.api

import io.github.pstanar.pstotp.core.model.api.VaultEntryUpsertResponse
import io.github.pstanar.pstotp.core.model.api.VaultSyncResponse
import org.json.JSONArray
import org.json.JSONObject

/** Vault sync API endpoints (authenticated — requires Bearer token). */
class VaultApi(private val client: ApiClient) {

    suspend fun fetchVault(): VaultSyncResponse =
        client.get("/vault") { VaultSyncResponse.fromJson(it) }

    suspend fun upsertEntry(
        entryId: String,
        entryPayload: String,
        entryVersion: Int,
    ): VaultEntryUpsertResponse {
        val body = JSONObject().apply {
            put("entryPayload", entryPayload)
            put("entryVersion", entryVersion)
        }
        return client.put("/vault/$entryId", body) { VaultEntryUpsertResponse.fromJson(it) }
    }

    suspend fun deleteEntry(entryId: String) {
        client.delete("/vault/$entryId")
    }

    suspend fun reorderEntries(entryIds: List<String>) {
        val body = JSONObject().apply {
            put("entryIds", JSONArray(entryIds))
        }
        client.post("/vault/reorder", body) { }
    }
}
